src/renderer/dx/CustomTextLayout.cpp

```diff
@@ -274,6 +288,39 @@ CustomTextLayout::CustomTextLayout(gsl::not_null<IDWriteFactory1*> const factory
             _glyphIndices.resize(totalGlyphsArrayCount);
         }
 
+        if (_isEntireTextSimple)
+        {
+            // When the entire text is simple, we can skip GetGlyphs and directly retrieve glyph indices and
+            // advances(in font design unit). With the help of font metrics, we can calculate the actual glyph
+            // advances without the need of GetGlyphPlacements. This shortcut will significantly reduce the time
+            // needed for text analysis.
+            DWRITE_FONT_METRICS1 metrics;
+            run.fontFace->GetMetrics(&metrics);
+
+            // With simple text, there's only one run. The actual glyph count is the same as textLength.
+            _glyphDesignUnitAdvances.resize(textLength);
+            _glyphAdvances.resize(textLength);
+            _glyphOffsets.resize(textLength);
+
+            USHORT designUnitsPerEm = metrics.designUnitsPerEm;
+
+            RETURN_IF_FAILED(_font->GetDesignGlyphAdvances(
+                textLength,
+                &_glyphIndices.at(glyphStart),
+                &_glyphDesignUnitAdvances.at(glyphStart),
+                run.isSideways));
+
+            for (size_t i = glyphStart; i < _glyphAdvances.size(); i++)
+            {
+                _glyphAdvances.at(i) = (float)_glyphDesignUnitAdvances.at(i) / designUnitsPerEm * _format->GetFontSize() * run.fontScale;
+            }
+
+            run.glyphCount = textLength;
+            glyphStart += textLength;
+
+            return S_OK;
+        }
+

```

Comment:
I'd prefer if we had this a little more orderly than an early-return block inside the existing method.

Would it be too much effort to make `_ShapeGlyphRun` check the `_isEntireTextSimple` at the top and then call one of two submethods:
1. `_ShapeSimpleGlyphRun()`
2. `_ShapeComplexGlyphRun()`

I think that'd be a bit more orderly.

Reply:
I think it’s a good idea. It’s just the right amount of effort.

Reply:
On second thought, I feel the need to break it into two methods is not very urgent. I found they share a lots of common variables like `textLength` and `glyphStart`. This method is very much like `AttrRow::InsertAttrRun` whose focus is on performance rather than readability.  Until we have some more complicate logic added, I feel it's justified to leave it here,.

Reply:
Alright. Thanks for considering.